website/src/pages/specification/20241027-flagsets.md

```diff
@@ -0,0 +1,200 @@
+- GO Feature Flag should have a `default` flag set for users not using the flag sets feature _(the behaviours should be exactly the same as of today if the feature is not used)_.
+- When calling the bulk evaluation APIs (`/ofrep/v1/evaluate/flags` or `/v1/allflags`), we will determine which flag set to evaluate based on API key used.
+- The bulk evaluation APIs should evaluate only 1 flag set at a time _(to avoid collision in flag names)_.
+- It should be able to specify which flag set to evaluate when calling the evaluation APIs. _(e.g. `/ofrep/v1/evaluate/flags` or `/ofrep/v1/evaluate/flags/{flag_key}`)_.
+  - Ideally we should be able to know which flag set to used based on the API Key used.
+  - If GOFF is configured to be public, we should be able to specify the flag set to evaluate in the request _(with a specific header ex:`goff-flag-set`)_.
+  - In the providers, we should be able to specify the flag set to evaluate directly in the constructor _(by providing an API Key, or the name of the flag set)_.
+- Admin API Keys, should be able to evaluate all the flag sets _(to be able to see all the flags available in the project)_. If none specified in the request, the default flag set should be evaluated.
+
+
+## Out of scope for now
+- Dynamic flag sets based on the evaluation context _([as mentioned in this slack message](https://gophers.slack.com/archives/C029TH8KDFG/p1732703075509229))_.
+- Single retrievers for multiple flag sets _(as proposed in https://github.com/thomaspoignant/go-feature-flag/issues/2314)_.
+
+_Even if out of scope for now, those are interesting options that we may want to implement later._
+
+## Proposed solution
+
+### Solution 1: 1 flag set per file
+
+In this solution we consider that we need at least one file per flag set.  
+All the flags retrieved by a `retriever` _(aka in the same file)_ will be associated to the same flag set.
+
+We can specify the flag set name in the configuration file.  
+As of today, we can still have multiple files for 1 flag set _(by specifying the same `flagSet` name in each file)_,
+and we will have the same mechanism as of today _(with flag overridden in case of flag name collision)_.
+
+#### Flags configuration file
+```yaml
+# config-file1.goff.yaml
+# Syntax used in this example is just for the sake of the example, it is not the final syntax.
+flagSet: flagset-teamA
+
+featureA-enabled:
+  variations:
+      enabled: true
+      disabled: false
+  defaultRule:
+    variation: enabled
+```
+```yaml
+# config-file2.goff.yaml
+flagSet: flagset-teamB
+
+featureA-enabled:
+  variations:
+      enabled: true
+      disabled: false
+  defaultRule:
+    variation: enabled
+```
+
+#### Relay-proxy configuration example
+```yaml
+# ...
+retrievers:
+  - kind: file
+    path: config-file1.goff.yaml
+  - kind: file
+    path: config-file2.goff.yaml
+
+authorizedKeys:
+  evaluation:
+    - apikey1 # owner: userID1
+    - apikey2 # owner: userID2
+  admin:
+    - apikey3
```

Comment:
🤔 I am not sure how to configure the relation between the API keys and the flag sets, all proposals are welcome on this.

Reply:
> I just went through the API keys configuration docs and it looks like it accept any string. Since it's an array it can be organized so that the available flag sets for a specific api key are declared as well in the config file.

Yes it can be anything here.
I am just not sure how to represent the configuration without breaking changes for existing config 🤔.

> Another approach would be to generate something like JWTs to be used as API keys and encode flag set access data as claims. This path would be much more complex
I was also thinking about that, but it could be an evolution for a later stage.